parl/framework/computation_wrapper.py

```diff
@@ -0,0 +1,161 @@
+    splits the batched results and sends them back to AgentHelper.
+    """
+
+    def __init__(self, name, ct, min_batchsize, max_batchsize, timeout,
+                 sample_method, **kwargs):
+        self.name = name
+        self.ct = ct
+        self.min_batchsize = min_batchsize
+        self.max_batchsize = max_batchsize
+        self.timeout = timeout
+        self.helper_creator = (
+            lambda comm: sample_method(name, comm, **kwargs))
+        self.comm = CommunicatorCT(self.timeout)
+        self.comms = {}
+        self.prediction_thread = Thread(target=self._prediction_loop)
+        self.training_thread = Thread(target=self._training_loop)
+        self.lock = Lock()
+        self.exit_flag = True
+        # TODO: get a parser of model specs 
+
+    def _pack_data(self, data):
+        """
+        Pack a list of data into one dict according to network's inputs specs.
+
+        Args:
+            data(list of dict): a list of data collected from Agents.
+        """
+        check_type_error(list, type(data))
+        starts = []
+        batch = {}
+        for k in data[0].iterkeys():
+            batch[k], t = concat_dicts((d[k] for d in data))
+            if not starts:
+                starts = t
+            else:
+                assert starts == t
+
+        return batch, starts
+
+    def _unpack_data(self, batch_dict_list, starts):
+        """
+        Unpack the dict into a list of dict, by slicing each value in the dict.        
+
+        Args:
+            batch_dict(dict): 
+        """
+        ret = []
+        for batch_dict in batch_dict_list:
+            dict_list = split_dict(batch_dict, starts)
+            assert not ret or len(ret[-1]) == len(dict_list)
+            ret.append(dict_list)
+
+        return zip(*ret)
+
+    def create_helper(self, agent_id):
+        comm = self.__create_communicator(agent_id)
+        return self.helper_creator(comm)
+
+    def __create_communicator(self, agent_id):
+        """
+        Creates a CommunicatorAgent (CA) with this ComputationWrapper (CW)'s
+        data communication channels (training and prediction Queues). Once an
+        AgentHelper of some Agent accepts this CA (i.e., the AgentHelper is
+        created with it), the Agent can exchange data with this CW through
+        this CA.
+        """
+        self.comms[agent_id] = CommunicatorAgent(
+            agent_id, self.comm.training_q, self.comm.prediction_q,
+            self.timeout)
+        return self.comms[agent_id]
+
+    def do_one_prediction(self, data):
+        data, starts = self._pack_data(data)
+        with self.lock:
+            ret = self.ct.predict(**data)
+        ret = self._unpack_data(ret, starts)
+        return ret
+
+    def do_one_training(self, data):
+        data, starts = self._pack_data(data)
+        with self.lock:
+            ret = self.ct.learn(**data)
+        #ret = self._unpack_data(ret, starts)
+        return ret
+
+    def _prediction_loop(self):
```

Comment:
Why do we have to do two communication to get training data ?
After each prediction, [state, action] can be stored in CT_communicator and used for training directly.
Currently, we predict the state and send it to agent.
Then, agent send data to CT and do training.
To generate an example for training, the same **state** was sent from agent twice currently.
Storing data in CT and use it for training would be a faster way.

Reply:
Sometimes, the data sent to do training are not the same as the data sent to do prediction. What needs to be done to the data before sending to do training is determined by the users through implementing run_one_episode. 

However, it is a good idea to support just sending data once.